File: cpp/cpp_100-999/757.cc
// C++
// leetcode 757
// https://leetcode.cn/problems/set-intersection-size-at-least-two/

class Solution {
public:
    static bool cmp(vector<int> &a, vector<int> &b) {
        if (a[1] == b[1]) {
            return b[0] - a[0] <= 0;
        }

        return a[1] - b[1] <= 0;
    }

    int intersectionSizeTwo(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end(), cmp);

        int a = -1, b = -1, res = 0;
        for (auto &i : intervals) {
            if (i[0] > b) {
                a = i[1] - 1;
                b = i[1];
                res += 2;
            } else if (i[0] > a) {
                a = b;
                b = i[1];
                res++;
            }
        }

        return res;
    }
};

File: Go/Go_100-999/757.go
// Go
// leetcode 757
// https://leetcode.cn/problems/set-intersection-size-at-least-two/

func intersectionSizeTwo(intervals [][]int) int {
	sort.Slice(intervals, func(i, j int) bool {
		if intervals[i][1] != intervals[j][1] {
			return intervals[i][1] < intervals[j][1]
		}
		return intervals[i][0] > intervals[j][0]
	})

	a, b, res := -1, -1, 0

	for i := 0; i < len(intervals); i++ {
		if intervals[i][0] > b {
			a = intervals[i][1] - 1
			b = intervals[i][1]
			res += 2
		} else if intervals[i][0] > a {
			a = b
			b = intervals[i][1]
			res++
		}
	}

	return res
}